File: problems/chessboard_and_queens/chessboard_and_queens.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<char>> board(8, vector<char>(8));
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            cin >> board[i][j];
        }
    }

    vector<bool> cols(8, false);
    vector<bool> diag1(15, false);
    vector<bool> diag2(15, false);

    int res = 0;

    auto dfs = [&](auto& self, int row) -> void {
        if (row == 8) {
            res++;
            return;
        }
        for (int col = 0; col < 8; col++) {
            if (board[row][col] == '*') continue;

            int d1 = row - col + 7;
            int d2 = row + col;

            if (cols[col] || diag1[d1] || diag2[d2]) continue;

            cols[col] = true;
            diag1[d1] = true;
            diag2[d2] = true;

            self(self, row + 1);

            cols[col] = false;
            diag1[d1] = false;
            diag2[d2] = false;
        }
    };

    dfs(dfs, 0);

    cout << res << "\n";

    return 0;
}

File: problems/CSES/Introductory_Problems/chessboard_and_queens/chessboard_and_queens.py
import sys
input = sys.stdin.readline

def main():
    board = [input() for _ in range(8)]

    cols = [False] * 8
    diag1 = [False] * 15
    diag2 = [False] * 15

    res = 0

    def dfs(row):
        nonlocal res

        # Если все 8 ферзей поставлены
        if row == 8:
            res += 1
            return

        # Пробуем поставить ферзя в каждый столбец
        for col in range(8):

            # Если запрещённая клетка
            if board[row][col] == "*":
                continue

            d1 = row - col + 7
            d2 = row + col

            # Если кто-то атакует
            if cols[col] or diag1[d1] or diag2[d2]:
                continue

            # Ставим ферзя
            cols[col] = True
            diag1[d1] = True
            diag2[d2] = True

            dfs(row + 1)

            # Убераем ферзя (backtracking)
            cols[col] = False
            diag1[d1] = False
            diag2[d2] = False

    dfs(0)

    print(res)

main()
